### ISSUE TITLE
[Documentation] Does LightGBM's C API make any guarantees on the dataset being immutable?

### ISSUE THREAD

**Issue by danielsoutar** [microsoft/LightGBM#5560]
I am working on using LightGBM as part of hyper-parameter search. For performance, I could either parallelise the training, which LightGBM offers (although I am working with the C API and through an immature FFI interface, so this isn't out-of-the-box functionality). Alternatively I could parallelise the search instead, which is easy to implement.

But I also need to ensure data race freedom in that case. It would be impractical to use a lock on the dataset handle between search threads, because that would render parallel search pointless and it would effectively just be sequential. If I don't use locks, then that means the library needs to guarantee that it doesn't modify the dataset through the handle when training/fitting. Is that the case? Can I make this assumption?

**Comment by danielsoutar** [comment]
Bumping this. Another point is that I could technically duplicate the dataset for each thread, but as you can imagine that is extremely memory-hungry and incurs very high overhead, so for real problems it would probably not be very useful.

**Comment by jameslamb** [comment]
@guolinke can you help with this question please?

**Comment by danielsoutar** [comment]
Hi again. This is quite important so doing some digging myself.

Looking through the C API, there is a `LGBM_DatasetSetField` call. This implies that the dataset can be mutated through the handle, but this isn’t necessarily an issue provided that multiple search threads set the same fields in advance before constructing their respective boosters. For instance, the `label` field would presumably be the same for each thread. I’m not sure what the `group` field is, but if it’s for ranking training that again is not an issue, nor weighted training. In all cases, this could be the same for each search thread. Only their model parameters need to differ.

The remaining ‘init_score’ field is unclear, but seems to refer to continued training (or effectively checkpointing) [here](https://lightgbm.readthedocs.io/en/latest/Parameters.html#continued-training-with-input-score). If that’s correct, then this could potentially be an issue in the general case, even assuming the same settings for everything else. This is because different models would score samples differently. I don’t think this is an issue in my case though because I could just disallow this kind of training checkpointing when doing parallel search.

In the call to `LGBM_BoosterAddValidData`, if applicable, the dataset passed is marked as `const Dataset*`. This means it is not mutated. The locking internally would then only apply to the booster, which is fine. Likewise when creating the booster:

```c++
// ---- start of booster

int LGBM_BoosterCreate(const DatasetHandle train_data,
                       const char* parameters,
                       BoosterHandle* out) {
  API_BEGIN();
  const Dataset* p_train_data = reinterpret_cast<const Dataset*>(train_data);
  auto ret = std::unique_ptr<Booster>(new Booster(p_train_data, parameters));
  *out = ret.release();
  API_END();
}
```

The actual object (not the pointer) is marked `const`, so this is safe.

The `Dataset` object does have a mutex in its declaration, which implies locking. This only applies to this function though, in `include/LightGBM/dataset.h`. I am not sure if this is an issue, or only relevant for initially loading the dataset:

```c++
/*! \brief Set whether the Dataset is finished automatically when last row is pushed or with a manual
 *         MarkFinished API call.  Set to true for thread-safe streaming and/or if will be coalesced later.
 *         FinishLoad should not be called on any Dataset that will be coalesced.
 */
inline void set_wait_for_manual_finish(bool value) {
    std::lock_guard<std::mutex> lock(mutex_);
    wait_for_manual_finish_ = value;
  }
```

Beyond that I can't see anything else. But again, a library maintainer would be a far better judge than me. Any response would be appreciated.

**Comment by danielsoutar** [comment]
To confirm: in the Booster definition in `src/c_api.cpp`, the Booster class has the `Dataset` marked as const:

```c++
class Booster {
// ...
 private:
  const Dataset* train_data_;
// ...
};
```

So my best guess would be that provided the dataset is fully prepared in the main thread, it can be shared between multiple boosters.

**Comment by jameslamb** [comment]
@guolinke @shiyu1994 (or maybe @AlbertoEAF ?) could one of you help with these questions?

**Comment by danielsoutar** [comment]
> Some changes of hyper parameters, like min_data_per_bin, min_data_per_leaf, max_bin, will cause the reconstruction of the dataset handle.

Thanks for the reply @guolinke. I'm assuming this potential reconstruction means it is not thread-safe in the general case? That's unfortunate we explore those parameters, which would mean I need to pass the pointer to the underlying memory buffer (which is fine) and build a `Dataset` object per thread in order to ensure thread-safety.

Looking at `LGBM_DatasetCreateFromMats`, I suspect that implies a lot of memory duplication if done per thread. So the answer to this question is probably 'No' for practical use in model search. Let me know if I'm misunderstanding/missing something.

**Comment by guolinke** [comment]
If one of the thread frees the handle (and then recreate), I think the remaining threads cannot know it, and thus is not thread-safe.
The simplest solution is to use a per-thread clone of Dataset object. 

BTW, you can refer to this function: https://github.com/microsoft/LightGBM/blob/f1d3181ced9fd01f4b2899054abd99be6773e939/python-package/lightgbm/basic.py#L2039-L2052 to know when the handle will be freed.

**Comment by danielsoutar** [comment]
Thanks for the help.
